### ISSUE TITLE
[`flake8-bugbear`] `B010`'s safe fix turns type-clean code into a type error when the attribute is a method

### ISSUE THREAD

**Issue by shawn-otto** [astral-sh/ruff#28493]
## Summary

`B010`'s fix is classified **safe**, but when the attribute being set is a **method**, applying it turns code that type-checks cleanly into a type error. It reproduces in both `ty` and `mypy`.

`setattr(obj, "method", fn)` is a common way to install a test double over a method precisely *because* the direct assignment doesn't type-check — so the fix removes the reason the `setattr` was written.

## Reproduction

```python
class Client:
    def _call(self, method: str) -> dict:
        return {}


def test_it() -> None:
    c = Client()

    def fake(method: str) -> dict:
        return {"ok": True}

    setattr(c, "_call", fake)
```

```console
$ ruff check --select B010 --fix repro.py
Found 1 error (1 fixed, 0 remaining).
```

No `--unsafe-fixes` needed. The last line becomes `c._call = fake`.

## Result

| | `ty 0.0.80` | `mypy 2.3.1` |
|---|---|---|
| `setattr(c, "_call", fake)` — before the fix | ✅ passes | ✅ passes |
| `c._call = fake` — after the fix | ❌ `invalid-assignment` | ❌ `error: Cannot assign to a method  [method-assign]` |

```console
$ mypy repro.py            # after
repro.py:12: error: Cannot assign to a method  [method-assign]
```

## Why this bites

In a project running both `ruff --fix` and a type checker in the same pre-commit hook set, the two undo each other on every run: ruff rewrites the `setattr` into an assignment, the type checker rejects the assignment, and there is no fixed point. The only forms that satisfy both are `cast(Any, fake)` or `# type: ignore`, i.e. suppressing the checker — which is strictly worse than the `setattr`, because it also stops the checker verifying the double's signature against the real method.

Assigning over a method is a long-standing type-checking limitation ([python/mypy#2427](https://github.com/python/mypy/issues/2427), open since 2016; [mypy's docs](https://mypy.readthedocs.io/en/stable/common_issues.html) suggest `Any` or `type: ignore`), so `setattr` is a natural escape and the rule removes it.

## Suggested resolution

Any of:

1. Don't flag `B010` when the attribute name resolves to a method on the target's type.
2. Keep the diagnostic, mark the fix **unsafe** in that case — matching the precedent in [#22657](https://github.com/astral-sh/ruff/pull/22657), where `B010`'s fix was made unsafe because it could delete comments.
3. Mark the fix unsafe whenever the assigned value is a callable.

(2) seems the smallest change consistent with existing behaviour.

## Version

`ruff 0.16.6`, `ty 0.0.80`, `mypy 2.3.1`, Python 3.13/3.14, macOS arm64.

**Comment by ntBre** [comment]
I'm open to other opinions, but this doesn't really fall under our typical fix safety criteria, which are that the fix doesn't change runtime behavior and doesn't remove comments. This is only changing the interaction with a type checker, which many Ruff users may not use.

I also think `setattr` is basically a black box for the type checker, meaning that it simply can't analyze it and find it problematic rather than analyzing it and "accepting" it as valid, if I understand correctly. This code also passes ty and mypy despite `fake` no longer having the correct signature:

```py
class Client:
    def _call(self, method: str) -> dict:
        return {}


def test_it() -> None:
    c = Client()

    def fake() -> dict:
        return {"ok": True}

    setattr(c, "_call", fake)
```

https://play.ty.dev/4ad37bd7-8e64-49ea-ac3e-1dfb097dcb29

I think something like [`unittest.mock.patch`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) would be the more direct way of accomplishing what you're doing than `setattr`, and that satisfies both Ruff and ty (although ty still doesn't verify that the signatures match with `patch`).

I'm hesitant to mark the fix as unsafe because it would require Ruff to infer the type of the object being assigned, which we can't do very reliably. Thus many fixes would likely have to become unsafe.

I'd be okay with documenting the limitation, though, as previously discussed in a similar context on https://github.com/astral-sh/ruff/issues/17263.

**Comment by GruffElixir** [comment]
Thanks, that makes sense. I’ve documented the type-checker limitation in #28509 and left the runtime fix-safety behavior unchanged: https://github.com/astral-sh/ruff/pull/28509
